lib/sedna/service/client.py

```diff
@@ -0,0 +1,218 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import json
+import time
+import threading
+from copy import deepcopy
+from retrying import retry
+from requests import request
+import websockets
+import asyncio
+from websockets.exceptions import InvalidStatusCode, WebSocketException, ConnectionClosedError, ConnectionClosedOK
+from sedna.common.log import sednaLogger
+
+
+@retry(stop_max_attempt_number=3, retry_on_result=lambda x: x is None, wait_fixed=2000)
+def http_request(url, method=None, timeout=None, binary=True, **kwargs):
+    _maxTimeout = timeout if timeout else 300
+    _method = "GET" if not method else method
+    try:
+        response = request(method=_method, url=url, **kwargs)
+        if response.status_code == 200:
+            return response.json() if binary else response.content.decode("utf-8")
+        elif 200 < response.status_code < 400:
+            sednaLogger.info(f"Redirect_URL: {response.url}")
+        sednaLogger.error('Get invalid status code %s while request %s', response.status_code, url)
+    except Exception as e:
+        sednaLogger.error(f'Error occurred while request {url}, Msg: {e}', exc_info=True)
+
+
+class LCReporter(threading.Thread):
+    """Inherited thread, which is an entity that periodically report to
+    the lc.
+    """
+
+    def __init__(self, message, period_interval=30):
+        threading.Thread.__init__(self)
+
+        # the value of statistics
+        self.inference_number = 0
+        self.hard_example_number = 0
+        self.period_interval = period_interval
+        # The system resets the period_increment after sending the messages to
+        # the LC. If the period_increment is 0 in the current period,
+        # the system does not send the messages to the LC.
+        self.period_increment = 0
+        self.message = deepcopy(message)
+        self.lock = threading.Lock()
+
+    def update_for_edge_inference(self):
+        self.lock.acquire()
+        self.inference_number += 1
+        self.period_increment += 1
+        self.lock.release()
+
+    def update_for_collaboration_inference(self):
+        self.lock.acquire()
+        self.inference_number += 1
+        self.hard_example_number += 1
+        self.period_increment += 1
+        self.lock.release()
+
+    def run(self):
+        while True:
+
+            start = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
+
+            time.sleep(self.period_interval)
+            if self.period_increment == 0:
+                continue
+
+            info = {
+                "startTime": start,
+                "updateTime": time.strftime("%Y-%m-%d %H:%M:%S", time.localtime()),
+                "inferenceNumber": self.inference_number,
+                "hardExampleNumber": self.hard_example_number,
+                "uploadCloudRatio": self.hard_example_number / self.inference_number
+            }
+            message = deepcopy(self.message)
```

Comment:
seems no need to deepcopy, remove it if unnecessary

Reply:
fix